spacelift/resource_role_attachment.go

```diff
@@ -0,0 +1,248 @@
+			},
+		},
+	}
+}
+
+func resourceRoleAttachmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiKeyID := d.Get("api_key_id").(string)
+
+	if apiKeyID != "" {
+		return createAPIKeyRoleBinding(ctx, d, meta)
+	}
+
+	return createIDPGroupMappingRoleBinding(ctx, d, meta)
+}
+
+func createAPIKeyRoleBinding(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var mutation struct {
+		CreateRoleBinding structs.APIKeyRoleBinding `graphql:"apiKeyRoleBindingCreate(input: $input)"`
+	}
+
+	variables := map[string]interface{}{
+		"input": structs.ApiKeyRoleBindingInput{
+			APIKeyID: toID(d.Get("api_key_id").(string)),
+			RoleID:   toID(d.Get("role_id").(string)),
+			SpaceID:  toID(d.Get("space_id").(string)),
+		},
+	}
+
+	if err := meta.(*internal.Client).Mutate(ctx, "ApiKeyRoleBindingCreate", &mutation, variables); err != nil {
+		return diag.Errorf("could not create role attachment: %v", internal.FromSpaceliftError(err))
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", apiKeyRoleAttachmentPrefix, mutation.CreateRoleBinding.ID))
+
+	return resourceRoleAttachmentRead(ctx, d, meta)
+}
+
+func createIDPGroupMappingRoleBinding(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var mutation struct {
+		CreateRoleBinding structs.UserGroupRoleBinding `graphql:"userGroupRoleBindingCreate(input: $input)"`
+	}
+
+	variables := map[string]interface{}{
+		"input": structs.UserGroupRoleBindingInput{
+			UserGroupID: toID(d.Get("idp_group_mapping_id").(string)),
+			RoleID:      toID(d.Get("role_id").(string)),
+			SpaceID:     toID(d.Get("space_id").(string)),
+		},
+	}
+
+	if err := meta.(*internal.Client).Mutate(ctx, "UserGroupRoleBindingCreate", &mutation, variables); err != nil {
+		return diag.Errorf("could not create user group role binding: %v", internal.FromSpaceliftError(err))
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", idpGroupMappingRoleAttachmentPrefix, mutation.CreateRoleBinding.ID))
+
+	return resourceRoleAttachmentRead(ctx, d, meta)
+}
+
+func resourceRoleAttachmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	id := d.Id()
+	if strings.HasPrefix(id, apiKeyRoleAttachmentPrefix) {
+		return readAPIKeyRoleBinding(ctx, d, meta)
+
+	}
+	return readIDPGroupMappingRoleBinding(ctx, d, meta)
+
+}
+
+func readAPIKeyRoleBinding(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var query struct {
+		APIKeyRoleBinding *structs.APIKeyRoleBinding `graphql:"apiKeyRoleBinding(id: $id)"`
+	}
+
+	id := strings.TrimPrefix(d.Id(), apiKeyRoleAttachmentPrefix+"/")
+	variables := map[string]interface{}{
+		"id": toID(id),
+	}
+
+	if err := meta.(*internal.Client).Query(ctx, "ApiKeyRoleBindingRead", &query, variables); err != nil {
+		return diag.Errorf("could not query for role attachment: %v", internal.FromSpaceliftError(err))
+	}
+
+	if query.APIKeyRoleBinding == nil {
```

Comment:
In what scenario can this happen? My understanding was that we either return a binding or an error. I'm just curious how can we not return an error and at the same time return an empty binding.